appserver/payara-appserver-modules/microprofile/metrics/src/main/java/fish/payara/microprofile/metrics/impl/ConcurrentGaugeImpl.java

```diff
@@ -95,59 +98,79 @@ public long getCount() {
 
     @Override
     public long getMax() {
-        clearOld();
-        long max = 0;
-        Instant nowMinStart = Instant.now().truncatedTo(ChronoUnit.MINUTES);
-        for (Map.Entry<Instant, Long> momentCount: lastCounts.entrySet()) {
-            if (momentCount.getKey().isBefore(nowMinStart) && momentCount.getValue() > max) {
-                max = momentCount.getValue();
-            }
-        }
-        return max;
+        currentStats();
+        return closedStats.max.get();
     }
 
     @Override
     public long getMin() {
-        clearOld();
-        long min = Long.MAX_VALUE;
-        // if it was not called in the last minute, then the value is 0
-        // but if it was called at least once in the last minute then the
-        //value is non-zero, even if it was 0 at some point duing that minute
-        if (lastCounts.isEmpty()) {
-            return 0;
-        }
-        Instant nowMinStart = Instant.now().truncatedTo(ChronoUnit.MINUTES);
-        for (Map.Entry<Instant, Long> momentCount: lastCounts.entrySet()) {
-            if (momentCount.getKey().isBefore(nowMinStart) && momentCount.getValue() < min) {
-                min = momentCount.getValue();
-            }
-        }
-        return min;
+        currentStats();
+        return closedStats.min.get();
     }
 
-    @Override
-    public void dec() {
-        clearOld();
-        lastCounts.put(Instant.now(), count.longValue());
-        count.decrement();
+    private static Instant getCurrentMinute() {
+        return Instant.now().truncatedTo(ChronoUnit.MINUTES);
+    }
+
+    private MinMax currentStats() {
+        MinMax possiblyOutdated = openStats.getAndUpdate(MinMax::replaceIfOld);
+        if (possiblyOutdated.finished) {
+            // we got previous MinMax instance, that has set finished=true just before it was replaced
+            closedStats = possiblyOutdated;
+            // if value was not updated for longer than one minute, this is still correct answer,
+            // as the gauge doesn't reset by itself.
+            return openStats.get();
+        } else {
+            return possiblyOutdated;
+        }
     }
-    
+
     /**
-     * Removes counts that occured before the previous minute that finished
+     * Stats captured by the gauge. Note that even if it is stored in
+     * AtomicReference, the class itself may be accessed concurrently.
      */
-    private void clearOld() {
-        Instant previousMinute = Instant.now().truncatedTo(ChronoUnit.MINUTES).minus(1, ChronoUnit.MINUTES);
-        if (previousMinute.equals(lastInstant)) {
-            //already called in this minute
-            return;
+    private class MinMax {
+
+        final AtomicLong min;
+        final AtomicLong max;
+        final Instant minute;
+        boolean finished;
```

Comment:
Since this is only relevant to the same thread setting it I am pretty sure this does not need any form of synchronisation. But its always tricky. A way to avoid the question altogether would be in `currentStats` first do `openStats.get()` and keep that in a local var, then do the `getAndUpdate`. If the update returns another `MinMax` instance it is outdated. The `finished` is no longer needed.

Reply:
`finished` is more optimization for the common case - usually the minute doesn't roll over, and then you only need to interact with atomic reference once.